enhancements/network/in-cluster-dns-and-loadbalancers-on-more-platforms.md

```diff
@@ -0,0 +1,297 @@
+  - TBD
+see-also:
+  - "/enhancements/network/baremetal-networking.md"
+replaces:
+superseded-by:
+---
+
+# In-cluster DNS and load balancers on more platforms
+
+## Summary
+
+Multiple on-prem platform types, including `baremetal` and `openstack`,
+[provide in-cluster network
+services](https://github.com/openshift/installer/blob/master/docs/design/baremetal/networking-infrastructure.md)
+that are required in order to have a viable multi-node cluster:
+
+* CoreDNS for in-cluster DNS resolution
+* haproxy with keepalived to provide in-cluster load balancers (ingress) for the API server and workloads
+
+Continuing the work from that original [enhancement
+proposal](https://github.com/openshift/enhancements/blob/master/enhancements/network/baremetal-networking.md),
+those services should also be available for optional inclusion when installing
+a cluster with the `external` or `none` platform types, which are likewise
+often used in environments that lack a suitable alternative for DNS and/or load
+balancers.
+
+## Motivation
+
+Provisioning and configuring a [DNS
+system](https://docs.openshift.com/container-platform/4.16/installing/installing_platform_agnostic/installing-platform-agnostic.html#installation-dns-user-infra_installing-platform-agnostic)
+and [load
+balancers](https://docs.openshift.com/container-platform/4.16/installing/installing_platform_agnostic/installing-platform-agnostic.html#installation-load-balancing-user-infra_installing-platform-agnostic)
+manually for an OpenShift cluster is a substantial burden on the user. In a
+cloud environment that's not already supported by the OpenShift installer,
+utilizing the native offerings requires additional work up-front, creates an
+ongoing maintenance burden, and more monetary cost for infrastructure. And not
+all cloud environments offer suitable options. On-prem, there may not exist
+sufficient DNS and/or load balancer services nor additional infrastructure on
+which to run them.
+
+Many users end up deploying a cluster with platform type `baremetal` when all
+they really want is to use the in-cluster network services. Instead, it should
+be possible to utilize those in-cluster network services without them being
+coupled to the baremetal platform type.
+
+For example, the assisted-installer and the agent based installer are often
+used to deploy clusters into “generic” environments where there is not an
+opportunity to utilize an external DNS or load balancer solution. Thus it sets
+the platform type as `baremetal` regardless of whether the systems are actually
+running on bare metal or whether there is any intent to use metal3
+integrations. The resulting cluster has all of the appearances of being bare
+metal, including a BareMetalHost resource for each Node, which can be confusing
+to users. Even the web console’s Overview landing page shows “Infrastructure
+provider: BareMetal” in addition to “N Nodes” and “N Bare Metal Hosts”.
+
```

Comment:
Unsure if we want to add it here, but this enhancement should mention Node OpenShift deployment (although its a bit distracting).
In SNO  (non-platform), DNS is handled independently using dnsmasq (when deployed by assisted). This proposal can align the behavior when installing SNO by enabling DNS service (load balancer ins't required) independently of the platform type.

Reply:
I assume "mention Node OpenShift" should say "mention Single Node OpenShift".

Is this what you're saying?

- since SNO uses platform type `none`, it typically requires the user to create the internal DNS records
- but assisted-installer alleviates that by configuring and enabling dnsmasq
- making dnsmasq enablement a native part of the `none` platform, as this proposal describes, would make that easier for all SNO users?

Reply:
Yes, that's great, need to find the right  place for it